torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -520,79 +522,50 @@ std::vector<at::IValue> readWrappedPayload(
-    } else if (
-        cudaProfilingEnabled && 0 == strcmp(e.name(), "__cuda_start_event")) {
-      e.setCudaUs(e.cpuUs());
-      auto device = e.device();
-      TORCH_CHECK(
-          device != -1,
-          "CUDA profiling was enabled but could not find CUDA device.");
-      TORCH_CHECK(
-          cudaProfilerStarts.find(device) == cudaProfilerStarts.end(),
-          c10::str("Duplicate __cuda_start_event found for ", device));
-      cudaProfilerStarts[device] = &e;
-    }
-
-    // TODO: determine no. of CUDA devices and break here if we have
-    // a cudaProfilerStart for all of them, in the case of cuda
-    // profiling.
-    if (foundCpuStart && !cudaProfilingEnabled) {
       break;
     }
   }
   // We should always find __start_profile.
   TORCH_CHECK(
       profilerStart != nullptr, "Expected to find __start_profile event.");
-  // Should have >= 1 CUDA start event if cudaProfilingEnabled.
-  // TODO: we can enhance this assert by ensuring we have found a
-  // start for every available CUDA device.
-  TORCH_CHECK(
-      !cudaProfilingEnabled || cudaProfilerStarts.size() > 0,
-      "Profiler was enabled with CUDA recording, but did not find __cuda_start_event.");
 
   if (cudaProfilingEnabled) {
-    // Compute and set global time for when this CUDA kernel was
-    // launched/ended, since deserialized event will not have a
-    // corresponding CUDA event.
+    // Deserialized events don't have the corresponding CUDA events, making it impossible
+    // to use cudaEventElapsedTime the receiving end. To avoid this, find all push/pop pairs
+    // of CUDA events and set the corresponding CUDA time to zero for the push event and
+    // to the elapsed time for the pop event, to be used later for the elapsed CUDA time
+    // computation.
+    std::unordered_map<at::RecordFunctionHandle, const LegacyEvent*> startEvents;
     for (auto& e : profiledEvents) {
       if (e.hasCuda()) {
-        auto cudaDevice = e.device();
-        TORCH_CHECK(
-            cudaDevice != -1,
-            "CUDA profiling was enabled but could not find CUDA device.");
-        auto it = cudaProfilerStarts.find(cudaDevice);
-        TORCH_CHECK(
-            it != cudaProfilerStarts.end(),
-            c10::str(
-                "Failed to find __cuda_start_event for device ", cudaDevice));
-        auto cudaProfilerStartEvent = it->second;
-        double cudaElapsedUs = cudaProfilerStartEvent->cudaElapsedUs(e);
-        int64_t cudaUs = cudaElapsedUs + cudaProfilerStartEvent->cpuUs();
-        e.setCudaUs(cudaUs);
+        if (e.kind() == EventKind::PushRange) {
+          startEvents[e.handle()] = &e;
+          e.setCudaUs(0);
+        } else if (e.kind() == EventKind::PopRange) {
+          auto it = startEvents.find(e.handle());
+          if (it != startEvents.end()) {
+            e.setCudaUs(it->second->cudaElapsedUs(e));
+          } else {
+            TORCH_WARN("Found a pop event without a corresponding push event");
```

Comment:
nit: might be good to log the event to help debugging. Also, should we `setCudaUs` to some reasonable value here?

Reply:
sg

Reply:
will set `setCudaUs`, also we don't have names in pop events